### ISSUE TITLE
Cool down heuristic

### ISSUE THREAD

**Issue by DreamLinuxer** [iu-parfunc/adaptive-data#4]
My current [adaptive bag](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs) still needs heuristic to determine whether the datastructure is cool down and should transition back to `PureBag`.

**Comment by rrnewton** [comment]
Yes, right now it looks a bit weird.  `transition` is called when `loop` tries `thresh` times without succeeding.  That means it's in "contented" status.  But then in transition it will actually do a backwards transition from `B -> BA`!  That seems like it will just make the data structure ping-pong back and forth between the two states...

**Comment by DreamLinuxer** [comment]
It won't, because once you go to the B state the `add` and `remove` just ignore the `thresh`. So once you goto B stata you will not go back unless you issue transition explicitly. (You can check [line 61](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs#L61) and [line 102](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs#L102))

**Comment by DreamLinuxer** [comment]
I think the type of `transition` should be change to `transition :: AdaptiveBag a -> Ticket (Hybrid a) -> IO ()` to prevent the in-flight add or remove to issue transition. [Fix](https://github.com/iu-parfunc/adaptive-data/commit/2d8263fdc9e1bb1790afa24976fb10330f326040)

**Comment by DreamLinuxer** [comment]
Failing scenario:
Current state is `A pbag 10`, the execution order:
Thread t1 executing [line 53 with `i==1`](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs#L53)
Thread t2 executing [line 121~135](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs#L121)
Thread t2 executing [line 93~101](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs#L93) until the bag is empty.
Thread t2 success in [line 100](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs#L100)
Thread t1 executing [line 45](https://github.com/iu-parfunc/adaptive-data/blob/3f05ef5e11c5ab1252e92418932e98f10267d5b1/chen_bag/src/AdaptiveBag.hs#L45) which will cause the transition from B -> BA.

`transition` should be called with the correct `Ticket`. [Fix](https://github.com/iu-parfunc/adaptive-data/commit/2d8263fdc9e1bb1790afa24976fb10330f326040)

**Comment by rrnewton** [comment]
Ah, right, the comment about ping-ponging was wrong.  But it is still the case that the latency hit is on the add/remove operation on the application threads.  Maybe that is ok, but it's does affect the asymptotic complexity of those ops.

**Comment by DreamLinuxer** [comment]
It won't affect the asymptotic complexity, in Scalablebag the asymptotic complexity of add and remove are both O(n).

**Comment by DreamLinuxer** [comment]
It only affect some add that issue transition, the complexity of remove remains O(n).

**Comment by rrnewton** [comment]
Wait a sec, with this simple list-based bag shouldn't the complexity be:

PureBag:
- `O(1)` add - cons on front of list
- `O(1)` remove - pop from front of list (just a stack)

"Scalable" bag:
- `O(1)` add - cons on front of thread-local list
- `O(P)` - may scan all thread-local lists for `P` processors.

Where would we get an `O(N)` where `N` is the size of the data structure?

**Comment by DreamLinuxer** [comment]
Oh, I just use n to denote the thread number, because the size of bag is irrelevant of our implementation.

**Comment by rrnewton** [comment]
Oh, right!  This version doesn't copy.

The one minor advantage of copying would be to "prefetch" elements from the purebag to the different per-thread slots.  But that is not worth it overall and could possibly be accomplished by one thread (that does the transition) doing a partial version of that -- scattering some elements but not all.

Alas, that's another thing that becomes less relevant with better data structures.  As far as I know, the good bag algorithms look like work-stealing structures where you aren't allowed to "push" into other thread's local areas.

**Comment by DreamLinuxer** [comment]
It is very hard to do copy while preserve set semantics(if you want to do this you need to lift every element in the list to EntryVal).
